File: apps/desktop/src-tauri/src/recents.rs
//! Recent-projects list + app-level prefs persisted in `app_config_dir`.
//!
//! Per `docs/workspace-redesign.md` Q7 the startup screen surfaces the last
//! 10 workspaces. Schema is intentionally tiny — path + display name + last
//! opened — and lives in `<app_config_dir>/recents.json` as plain JSON so a
//! user can hand-edit / reset by deleting the file.
//!
//! "Reopen last project on launch" (Q7 sub-decision, default off) lives in
//! the same file so the startup screen doesn't have to read two files.
//! It's deliberately opt-in: a partial-commit state at crash time should not
//! silently re-mount on the next boot.

use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Arc, RwLock};

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tracing::warn;

const RECENTS_FILE: &str = "recents.json";
const MAX_RECENTS: usize = 10;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct RecentEntry {
    pub path: PathBuf,
    pub name: String,
    pub last_opened: DateTime<Utc>,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
struct RecentsFile {
    #[serde(default)]
    reopen_on_launch: bool,
    #[serde(default)]
    entries: Vec<RecentEntry>,
    /// Parent folder of the last project the user created via the
    /// "+ New project" form. The startup screen pre-fills this so the
    /// next new project lands next to the previous one without
    /// re-navigating from `C:\Users\<name>\`. Falls back to the OS
    /// Documents directory in the UI when this is `None` (first launch).
    #[serde(default)]
    last_new_project_parent: Option<PathBuf>,
}

/// Tauri-managed recents store. Holds the path to `recents.json` and
/// serialises reads/writes through an RwLock; per-startup-screen access
/// cost is one file read + JSON parse, so caching isn't worth the
/// invalidation complexity.
#[derive(Clone)]
pub struct RecentsStore {
    path: Arc<RwLock<PathBuf>>,
}

impl RecentsStore {
    /// Create a store rooted at `<app_config_dir>/recents.json`. Caller is
    /// responsible for ensuring `config_dir` exists.
    pub fn new(config_dir: PathBuf) -> Self {
        Self {
            path: Arc::new(RwLock::new(config_dir.join(RECENTS_FILE))),
        }
    }

    fn read(&self) -> Result<RecentsFile> {
        let path = self.path.read().expect("recents path lock").clone();
        if !path.exists() {
            return Ok(RecentsFile::default());
        }
        let body = fs::read_to_string(&path)
            .with_context(|| format!("read {}", path.display()))?;
        if body.trim().is_empty() {
            return Ok(RecentsFile::default());
        }
        serde_json::from_str(&body).with_context(|| format!("parse {}", path.display()))
    }

    fn write(&self, file: &RecentsFile) -> Result<()> {
        let path = self.path.read().expect("recents path lock").clone();
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)
                .with_context(|| format!("create {}", parent.display()))?;
        }
        let json = serde_json::to_string_pretty(file).context("serialize recents")?;
        let tmp = path.with_extension("json.tmp");
        fs::write(&tmp, json).with_context(|| format!("write {}", tmp.display()))?;
        fs::rename(&tmp, &path)
            .with_context(|| format!("promote {} -> {}", tmp.display(), path.display()))?;
        Ok(())
    }

    /// Push `(path, name)` to the top of the list, deduping by canonical
    /// path (case-insensitive on Windows is good enough — the user picked
    /// the same folder). Truncates to `MAX_RECENTS`. Idempotent on retry;
    /// a failed write is logged but doesn't propagate to the caller —
    /// recents is best-effort UX, not a correctness anchor.
    pub fn push(&self, path: PathBuf, name: String) {
        let now = Utc::now();
        let mut file = match self.read() {
            Ok(f) => f,
            Err(e) => {
                warn!("recents read failed, starting fresh: {e:#}");
                RecentsFile::default()
            }
        };
        file.entries.retain(|e| !same_path(&e.path, &path));
        file.entries.insert(
            0,
            RecentEntry {
                path,
                name,
                last_opened: now,
            },
        );
        file.entries.truncate(MAX_RECENTS);
        if let Err(e) = self.write(&file) {
            warn!("recents write failed: {e:#}");
        }
    }

    /// Remove an entry by path. No-op if the path isn't in the list.
    pub fn remove(&self, path: &Path) -> Result<()> {
        let mut file = self.read()?;
        file.entries.retain(|e| !same_path(&e.path, path));
        self.write(&file)
    }

    /// Read the current list, newest first.
    pub fn list(&self) -> Result<Vec<RecentEntry>> {
        Ok(self.read()?.entries)
    }

    /// Top entry, if any. Used by the "Reopen last project on launch" path.
    pub fn most_recent(&self) -> Result<Option<RecentEntry>> {
        Ok(self.read()?.entries.into_iter().next())
    }

    pub fn reopen_on_launch(&self) -> Result<bool> {
        Ok(self.read()?.reopen_on_launch)
    }

    pub fn set_reopen_on_launch(&self, value: bool) -> Result<()> {
        let mut file = self.read().unwrap_or_default();
        if file.reopen_on_launch == value {
            return Ok(());
        }
        file.reopen_on_launch = value;
        self.write(&file)
    }

    /// Last parent folder used in the "+ New project" form. The startup
    /// screen pre-fills "Save in" with this so the user doesn't have to
    /// re-navigate from the OS root every time. `None` on first launch —
    /// the UI falls back to the OS Documents directory.
    pub fn last_new_project_parent(&self) -> Result<Option<PathBuf>> {
        Ok(self.read()?.last_new_project_parent)
    }

    /// Record the parent folder of the just-created workspace. Called by
    /// `project_new_workspace` on success. Best-effort: any error is
    /// logged but doesn't propagate to the caller — losing the default
    /// folder is mild UX regression, not a correctness anchor.
    pub fn set_last_new_project_parent(&self, parent: PathBuf) {
        let mut file = match self.read() {
            Ok(f) => f,
            Err(e) => {
                warn!("recents read failed, starting fresh: {e:#}");
                RecentsFile::default()
            }
        };
        file.last_new_project_parent = Some(parent);
        if let Err(e) = self.write(&file) {
            warn!("recents write failed: {e:#}");
        }
    }
}

#[cfg(target_os = "windows")]
fn same_path(a: &Path, b: &Path) -> bool {
    a.to_string_lossy()
        .eq_ignore_ascii_case(&b.to_string_lossy())
}

#[cfg(not(target_os = "windows"))]
fn same_path(a: &Path, b: &Path) -> bool {
    a == b
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn fresh(tmp: &TempDir) -> RecentsStore {
        RecentsStore::new(tmp.path().to_path_buf())
    }

    #[test]
    fn empty_when_no_file_yet() {
        let tmp = TempDir::new().unwrap();
        let store = fresh(&tmp);
        assert!(store.list().unwrap().is_empty());
        assert!(store.most_recent().unwrap().is_none());
        assert!(!store.reopen_on_launch().unwrap());
    }

    #[test]
    fn push_dedupes_and_caps() {
        let tmp = TempDir::new().unwrap();
        let store = fresh(&tmp);
        for i in 0..15 {
            store.push(PathBuf::from(format!("/proj/p{i}")), format!("p{i}"));
        }
        let entries = store.list().unwrap();
        assert_eq!(entries.len(), MAX_RECENTS);
        // Most recent first.
        assert_eq!(entries[0].name, "p14");

        // Re-push an existing entry — should move to top, not duplicate.
        store.push(PathBuf::from("/proj/p10"), "p10".into());
        let entries = store.list().unwrap();
        assert_eq!(entries.len(), MAX_RECENTS);
        assert_eq!(entries[0].name, "p10");
    }

    #[test]
    fn remove_drops_entry() {
        let tmp = TempDir::new().unwrap();
        let store = fresh(&tmp);
        store.push(PathBuf::from("/proj/a"), "a".into());
        store.push(PathBuf::from("/proj/b"), "b".into());
        store.remove(Path::new("/proj/a")).unwrap();
        let entries = store.list().unwrap();
        assert_eq!(entries.len(), 1);
        assert_eq!(entries[0].name, "b");
    }

    #[test]
    fn reopen_on_launch_round_trips() {
        let tmp = TempDir::new().unwrap();
        let store = fresh(&tmp);
        assert!(!store.reopen_on_launch().unwrap());
        store.set_reopen_on_launch(true).unwrap();
        assert!(store.reopen_on_launch().unwrap());
        store.set_reopen_on_launch(false).unwrap();
        assert!(!store.reopen_on_launch().unwrap());
    }

    #[test]
    fn last_new_project_parent_round_trips() {
        let tmp = TempDir::new().unwrap();
        let store = fresh(&tmp);
        assert!(store.last_new_project_parent().unwrap().is_none());
        store.set_last_new_project_parent(PathBuf::from("/projects/area"));
        assert_eq!(
            store.last_new_project_parent().unwrap(),
            Some(PathBuf::from("/projects/area")),
        );
        // Overwrite — UI re-records on every new project.
        store.set_last_new_project_parent(PathBuf::from("/other/area"));
        assert_eq!(
            store.last_new_project_parent().unwrap(),
            Some(PathBuf::from("/other/area")),
        );
    }
}

File: apps/desktop/src/main/recents.ts
// Recent-projects list + startup prefs persisted at <userData>/recents.json,
// owned by the Electron main process (config-dir, one value across every
// project).
//
// History: persistence used to live in the Rust addon (native/src/recents.rs);
// it moved here so the addon is compute-only and the Rust↔TS struct twin is
// gone. The on-disk file path + JSON field names are unchanged, so existing
// users' recents.json keeps working after the move to TS.
//
// On-disk shape:
//   {
//     "reopen_on_launch": bool,
//     "entries": [ { "path": "…", "name": "…", "last_opened": "<ISO-8601>" }, … ],
//     "last_new_project_parent": "…" | null
//   }
//
// Bad-config recovery: a missing / empty / corrupt file degrades to all-defaults
// (entries=[], reopen_on_launch=false, last_new_project_parent=null) so a hand-
// edit mishap can't brick the editor (parity with the old Rust store).
//
// push() and setLastNewProjectParent() are BEST-EFFORT: they log and swallow on
// any fs/parse error (parity with the Rust store's warn+return pattern).
//
// No :changed event (parity with Rust behavior — renderer re-fetches via channel
// calls as needed).

import type { RecentEntry } from '../shared/recents'

const MAX_RECENTS = 10

/** Minimal fs surface — injected so tests run in-memory; node:fs in production. */
export interface RecentsFs {
  exists(path: string): boolean
  readFile(path: string): string
  writeFile(path: string, text: string): void
  rename(from: string, to: string): void
  mkdirp(dir: string): void
}

export interface RecentsStore {
  /** Read the current list, newest first. */
  list(): RecentEntry[]
  /** Remove an entry by path. No-op if the path isn't in the list. */
  remove(path: string): void
  /** Top entry, if any. Used by the "Reopen last project on launch" path. */
  mostRecent(): RecentEntry | null
  getReopenOnLaunch(): boolean
  setReopenOnLaunch(value: boolean): void
  lastNewProjectParent(): string | null
  /** Push (path, displayName) to top, dedupe by path (case-insensitive on Windows),
   *  cap at 10, stamp last_opened = new Date().toISOString().
   *  BEST-EFFORT: never throws (log + swallow on any fs/parse error). */
  push(path: string, displayName: string): void
  /** Record the parent folder of the just-created workspace.
   *  BEST-EFFORT: never throws (log + swallow on any fs/parse error). */
  setLastNewProjectParent(parent: string): void
}

/** On-disk envelope. */
interface RecentsFile {
  reopen_on_launch: boolean
  entries: RecentEntry[]
  last_new_project_parent: string | null
}

const DEFAULTS: RecentsFile = { reopen_on_launch: false, entries: [], last_new_project_parent: null }

/** Normalize path for dedup comparison, mirroring the old Rust `same_path`
 *  (#[cfg]-split): case-INSENSITIVE on Windows (lower-cased), case-SENSITIVE
 *  everywhere else (returned verbatim). This preserves exact parity with the
 *  Rust store across the 3-OS matrix — on macOS/Linux `/Proj/A` and `/proj/a`
 *  are DISTINCT recents, on Windows they collapse to one. */
function normPath(p: string): string {
  return process.platform === 'win32' ? p.toLowerCase() : p
}

export function createRecentsStore(deps: { fs: RecentsFs; path: string; dir: string }): RecentsStore {
  function read(): RecentsFile {
    if (!deps.fs.exists(deps.path)) return { ...DEFAULTS, entries: [] }
    let body: string
    try { body = deps.fs.readFile(deps.path) }
    catch (e) { console.warn(`[recents] read ${deps.path}:`, e); return { ...DEFAULTS, entries: [] } }
    if (body.trim() === '') return { ...DEFAULTS, entries: [] }
    try {
      const parsed = JSON.parse(body) as Partial<RecentsFile>
      return {
        reopen_on_launch: typeof parsed.reopen_on_launch === 'boolean' ? parsed.reopen_on_launch : false,
        entries: Array.isArray(parsed.entries) ? (parsed.entries as RecentEntry[]) : [],
        last_new_project_parent: typeof parsed.last_new_project_parent === 'string' ? parsed.last_new_project_parent : null,
      }
    } catch (e) {
      console.warn(`[recents] parse ${deps.path}:`, e); return { ...DEFAULTS, entries: [] }
    }
  }

  function write(file: RecentsFile): void {
    deps.fs.mkdirp(deps.dir)
    const tmp = deps.path + '.tmp'
    deps.fs.writeFile(tmp, JSON.stringify(file, null, 2))
    deps.fs.rename(tmp, deps.path) // atomic promote
  }

  return {
    list(): RecentEntry[] {
      return read().entries
    },

    remove(path: string): void {
      const file = read()
      const norm = normPath(path)
      file.entries = file.entries.filter((e) => normPath(e.path) !== norm)
      write(file)
    },

    mostRecent(): RecentEntry | null {
      return read().entries[0] ?? null
    },

    getReopenOnLaunch(): boolean {
      return read().reopen_on_launch
    },

    setReopenOnLaunch(value: boolean): void {
      const file = read()
      if (file.reopen_on_launch === value) return
      file.reopen_on_launch = value
      write(file)
    },

    lastNewProjectParent(): string | null {
      return read().last_new_project_parent
    },

    push(path: string, displayName: string): void {
      const now = new Date().toISOString()
      let file: RecentsFile
      try { file = read() }
      catch (e) { console.warn('[recents] read failed on push, starting fresh:', e); file = { ...DEFAULTS, entries: [] } }
      const norm = normPath(path)
      file.entries = file.entries.filter((e) => normPath(e.path) !== norm)
      file.entries.unshift({ path, name: displayName, last_opened: now })
      if (file.entries.length > MAX_RECENTS) file.entries.length = MAX_RECENTS
      try { write(file) }
      catch (e) { console.warn('[recents] write failed on push:', e) }
    },

    setLastNewProjectParent(parent: string): void {
      let file: RecentsFile
      try { file = read() }
      catch (e) { console.warn('[recents] read failed on setLastNewProjectParent, starting fresh:', e); file = { ...DEFAULTS, entries: [] } }
      file.last_new_project_parent = parent
      try { write(file) }
      catch (e) { console.warn('[recents] write failed on setLastNewProjectParent:', e) }
    },
  }
}
